sqlglot/dialects/snowflake.py

```diff
@@ -542,6 +542,41 @@ def _annotate_reverse(self: TypeAnnotator, expression: exp.Reverse) -> exp.Rever
     return expression
 
 
+def _annotate_decode_case(self: TypeAnnotator, expression: exp.DecodeCase) -> exp.DecodeCase:
+    """Annotate DecodeCase with the type inferred from return values only.
+
+    DECODE uses the format: DECODE(expr, val1, ret1, val2, ret2, ..., default)
+    We only look at the return values (ret1, ret2, ..., default) to determine the type,
+    not the comparison values (val1, val2, ...) or the expression being compared.
+    """
+    self._annotate_args(expression)
+
+    expressions = expression.expressions
+    if not expressions:
+        self._set_type(expression, exp.DataType.Type.UNKNOWN)
+        return expression
+
+    # Return values are at indices 2, 4, 6, ... and the last element (if odd length)
```

Comment:
This should be "if even length". For example if you have `DECODE(expr, k, v, default)`, that's 4 args. I think it was fixed a few lines below as well (L563).